### PR TITLE
Add new @syms syntax

### PR DESCRIPTION
**Pull request by matthieubulte**
See #416 

I just noticed the `@symfuns` macro which has the syntax of `@vars`, this is may be confusing as well. Would you like me to do something about it?

### CODE REVIEW

**Comment by jverzani** [comment]
That is, if I recall, an alternative to the `SymFunction`. If you could have some parseable syntax to indicate that `u` is supposed to be a function and not a variable, that would be *awesome*. (I would hope, maybe something like `u()` would be enough?)

**Comment by matthieubulte** [comment]
@jverzani yes `u()` was enough. You can check out the unit tests to see how it interacts with the other features.

File: test/tests.jl [hunk 1]
```diff
@@ -487,8 +487,33 @@ import PyCall
...
+    
+    @test isa(y, Sym)
+
+    @test isa(z, Sym)
+    @test ask(𝑄.complex(z))
+    
+    @test isa(n, Sym)
+    @test ask(𝑄.integer(n))
+
+    f, g, h = @syms f()::(real, positive), g(), h()::complex=>"h̄"
```
Comment (jverzani) [inline test/tests.jl:504]: Wow, I hadn't known you could annotate the output. This is really nice.

**Comment by jverzani** [comment]
Let me run the CI tests. If they pass, is this ready to merge? I'm really thankful for this, it will be much nicer for the users.

File: test/tests.jl (hunk 1 above)
Comment (matthieubulte) [reply test/tests.jl:504]: I was surprised as well, I found this in the unit tests for `@symfuns`.

**Comment by matthieubulte** [comment]
>  If they pass, is this ready to merge?

I think so, yes.

**Comment by jverzani** [comment]
I was starting to document this, and found this simple command isn't processed quite right:

```
julia> @syms a
ERROR: LoadError: type Symbol has no field head
```

**Comment by matthieubulte** [comment]
Oops, sorry about that. Fortunately, the fix is easy.

At https://github.com/JuliaPy/SymPy.jl/blob/master/src/constructors.jl#L48
```julia
if length(xs) == 1 && xs[1].head == :tuple
```
should be
```julia
if length(xs) == 1 && isa(xs[1], Expr) && xs[1].head == :tuple
```

**Comment by jverzani** [comment]
I got, thanks! (Boy you work fast!)

Another thing I noticed, this design needs to assign the variables, right? (as in `a,b = @syms a,b`) Is there a way to skirt around the duplication of values (e.g., just `@syms a,b`?)

**Comment by jverzani** [comment]
Actually, it seems `@syms V` creates `V`, but `@syms U,V` needs the assignment.

**Comment by jverzani** [comment]
Wait, sorry, this works `@syms U1 U2 U3`, but this doesn't for this purpose `@syms U1, U2, U3`. Should I document that distinction, or is it something to adjust while this is fresh?

**Comment by matthieubulte** [comment]
Hmm, that's weird. I failed to reproduce the issue of variables not getting defined, but maybe it's from the way I use julia. There seems to be a difference in the macro expansion of both indeed, let me see where it might come from.

**Comment by jverzani** [comment]
Here is what I see:

```
julia> U1
ERROR: UndefVarError: U1 not defined

julia> @syms U1 U2 U3
(U1, U2, U3)

julia> U1
U₁

julia> W1
ERROR: UndefVarError: W1 not defined

julia> @syms W1, W2, W3
(W1, W2, W3)

julia> W1
ERROR: UndefVarError: W1 not defined
```

Of course, this last example makes me ask, would supporting

```
@syms U[1:3]
```

to define U1, U2, U3, be easy to add?

**Comment by matthieubulte** [comment]
Ok, so here is a possible fix
```julia

macro syms(xs...)
    # If the user separates declaration with commas, the top-level expression is a tuple
    if length(xs) == 1 && isa(xs[1], Expr) && xs[1].head == :tuple
        _gensyms(xs[1].args...)
    elseif length(xs) > 0
        _gensyms(xs...)
    end
end

function _gensyms(xs...)
    asstokw(a) = Expr(:kw, esc(a), true)
    
    # Each declaration is parsed and generates a declaration using `symbols`
    symdefs = map(xs) do expr
        varname, sym, assumptions, isfun = parsedecl(expr)
        ctor = isfun ? :SymFunction : :symbols
        sym, :($(esc(sym)) = $(ctor)($(varname), $(map(asstokw, assumptions)...)))
    end
    syms, defs = collect(zip(symdefs...))

    # The macro returns a tuple of Symbols that were declared
    Expr(:block, defs..., :(tuple($(map(esc,syms)...))))
end
```

I'm not super happy, but I'm not sure how to keep using only the macro without having issues with julia escaping in a weird way the variables.

For ref, the previous implementation would do the following:
```julia
julia> @macroexpand @syms a7, a8, a9
quote
    Main.a7 = Main.symbols("a7")
    Main.a8 = Main.symbols("a8")
    Main.a9 = Main.symbols("a9")
    Main.tuple(Main.a7, Main.a8, Main.a9)
end
```
which, depending on how you run the code, might use the wrong module as a prefix. Without the commas, this was always fine as you noted
```julia
julia> @macroexpand @syms a7 a8 a9
quote
    a7 = Main.symbols("a7")
    a8 = Main.symbols("a8")
    a9 = Main.symbols("a9")
    Main.tuple(a7, a8, a9)
end
```

The new implementation (which remove the recursive macro application) just works.

> Of course, this last example makes me ask, would supporting
> 
> @syms U[1:3]
> to define U1, U2, U3, be easy to add?

Yes, this is easy to do. I was about to start when I remembered that I have other things a little more urgent to do at the moment and that I should stop procrastinating. I can come back with this in the near future.

**Comment by jverzani** [comment]
Thanks again! Do I need to unmerge or something to allow you to modify the merged PR?

**Comment by matthieubulte** [comment]
well, thank you for finding these bugs :) I answered your question in the new PR, but on my side no need to unmerge. For the users, I guess as long as you don't tag/release, it's "ok" to temporarily have a broken master.

**Pull request merged**